xmr-btc/tests/harness/storage.rs

```diff
@@ -0,0 +1,158 @@
+use anyhow::{anyhow, Context, Result};
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use std::path::Path;
+
+pub struct Database<T>
+where
+    T: Serialize + DeserializeOwned,
+{
+    db: sled::Db,
+    _marker: std::marker::PhantomData<T>,
+}
+
+impl<T> Database<T>
+where
+    T: Serialize + DeserializeOwned,
+{
+    const LAST_STATE_KEY: &'static str = "latest_state";
+
+    pub fn open(path: &Path) -> Result<Self> {
+        let db =
+            sled::open(path).with_context(|| format!("Could not open the DB at {:?}", path))?;
+
+        Ok(Database {
+            db,
+            _marker: Default::default(),
+        })
+    }
+
+    pub async fn insert_latest_state(&self, state: &T) -> Result<()> {
+        let key = serialize(&Self::LAST_STATE_KEY)?;
+        let new_value = serialize(&state).context("Could not serialize new state value")?;
+
+        let old_value = self.db.get(&key)?;
+
+        self.db
+            .compare_and_swap(key, old_value, Some(new_value))
+            .context("Could not write in the DB")?
+            .context("Stored swap somehow changed, aborting saving")?;
+
+        self.db
+            .flush_async()
```

Comment:
Do note that sled automatically flushes in the background several time per second (it's configurable) so adding a todo to rethink the usage of `flush_async` is likely to help performance once done.

Reply:
added a todo